### PR TITLE
Count distinct returned incorrect results without useApproximateCountDistinct

### PR DESCRIPTION
**Pull request by kgyrtkirk**
With `useApproximateCountDistinct=false` queries like:
```
select count(distinct m1) from druid.foo where m1 < -1.0
```
may have returned incorrected results.

This PR has:

- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregatorFactory.java [hunk 1]
```diff
@@ -85,6 +85,12 @@ public final int getMaxIntermediateSizeWithNulls()
     return getMaxIntermediateSize() + (NullHandling.replaceWithDefault() ? 0 : Byte.BYTES);
   }
 
+  protected boolean canHandleNulls()
+  {
+    return NullHandling.replaceWithDefault();
+  }
```
Comment (kgyrtkirk) [inline processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregatorFactory.java:91]: @rohangarg @clintropolis: how do you guys see this api change?
* it could be interpreted as something which is in conflict with the original intention of the interface
  * as it declares in the APIdoc that it will be `null` if nothing is aggregated
* but on the otherhand it somewhat aligns with the original intention as well to give more fine-grained control to the  underlying aggregator on how it would like to handle the `null` values

long story short: the situation here is that a `Long` aggregtor is needed which results in `0` in case of no rows being aggregated - especially in case `useDefaultValueForNull` is turned off

**Comment by kgyrtkirk** [comment]
2 checks have hit the nodejs download issue of #14861 - all other checks have passed!
@clintropolis, @rohangarg  could you please take a look?

File: processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregatorFactory.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregatorFactory.java:91]: This reminds me of an old PR of mine which I totally forgot about until reviewing this PR, #10666, which made a different but somewhat similar change.

In that PR, I was adding a method called `hasNulls` to the nullable numeric aggregator (side note, i probably should resurrect this old PR), to separate processing in SQL compatible mode for columns which have nulls and those which don't, but which should still output a null value if not aggregated.

The changes in that PR would not actually fix the problems here, and I think maybe help me frame what actually bothers me about the changes in this PR, since the changes here are somewhat contradictory to the changes there.

Anyway, I think what bothers me about this change is that it is to handle logic that seems entirely specific to the special case of how COUNT aggregations are handled, and I can't actually see any good use for this otherwise, and because of that I'm not entirely sure it makes sense to add in this way. If we ignore the default value mode (which is now a legacy mode as of #14792), and consider the changes i made in the other PR, then this family of aggregators is for implementing math aggregator functions which may or may not have null values, but should always be null unless aggregated.

COUNT however, is a special case in that it should be 0, even if not aggregated, so I think adding special handling to this family aggregators specifically for the 'merging' aggregator for COUNT seems like the wrong thing to do.

Instead, I think i would be more in favor of just special case implementing the merging/combining aggregator for COUNT, even if it means not sharing some of the same code that we use for SUM, because it is a special case that i can't see having a general use for other types of numeric aggregators, only COUNT.

Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregatorFactory.java:91]: Yes; actually - I wanted a way to declare that this is `NullAware` :D 

Earlier I kinda wanted to take a similar path as you suggested  - just alter the combiner - but that ended up as this change eventually; to my knowledge:

* `CombiningSequence` creates the accumulator and dispatches it thru `Sequence#accumulate`
* as per what I've seen/remember the *first* element is not run thru the combiner as there is nothing to combine it too... 
* there is only a [combine](https://github.com/apache/druid/blob/194a9c9abc1b6edf98c2652687515200c6cbd3e8/processing/src/main/java/org/apache/druid/query/aggregation/CountAggregatorFactory.java#L85) method which has 2 arguments
* or the combining factory...

actually its interesting to consider a corner case (I don't have a testcase for this - but it was covered by existing cases):

* `CombiningSequence` is run on an inputstream which produces only 1 output element - which is produced via the `AggregateCominer` from a number of input
* The `AggregateCombiner` for `CountAggregatorFactory` is `LongSumAggregateCombiner` originally - so it may have returned null ..and thus by giving a sole input value could skip the combine method call

I was considering at some point also to extend the `Accumulator` interface so that the it could communicate the initial value 

Will investigate your suggestion/other alternatives - but iirc when there are 0 elements aggregated those methods are not called/work....that's why I resorted to the current solution.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryRunnerFactory.java [hunk 2]
```diff
@@ -108,4 +124,57 @@ public GroupingEngine getGroupingEngine()
...
+  }
+
+  private static Iterator<ResultRow> summaryRowIterator(GroupByQuery q)
+  {
+    List<AggregatorFactory> aggSpec = q.getAggregatorSpecs();
+    Object[] values = new Object[aggSpec.size()];
+    for (int i = 0; i < aggSpec.size(); i++) {
+      values[i] = aggSpec.get(i).factorize(new AllNullColumnSelectorFactory()).get();
+    }
+    return Collections.singleton(ResultRow.of(values)).iterator();
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryRunnerFactory.java:177]: There's a `Collections.singletonIterator` that you can use instead.  It's a nit, but will save on an object allocation.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java [hunk 3]
```diff
@@ -175,10 +175,13 @@ private Sequence<ResultRow> mergeGroupByResults(
...
   {
+    Sequence<ResultRow> process;
     if (isNestedQueryPushDown(query)) {
-      return mergeResultsWithNestedQueryPushDown(query, resource, runner, context);
+      process = mergeResultsWithNestedQueryPushDown(query, resource, runner, context);
+    } else {
+      process = mergeGroupByResultsWithoutPushDown(query, resource, runner, context);
     }
-    return mergeGroupByResultsWithoutPushDown(query, resource, runner, context);
+    return GroupByQueryRunnerFactory.wrapSummaryRowIfNeeded(query, process);
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java:184]: I'm surprised that this was required, which test caused you to need this change?  I say this because the only way you should be able to get a completely empty sequence here is if the "leaf nodes" are producing completely empty sequences.  The change in the other place should ensure that no leaf node ever produces a completely empty sequence, meaning that this change shouldn't be necessary...

**Review by imply-cheddar — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java (hunk 3 above)
Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java]: Thank you for taking a look!
unfortunately its needed - I've linked the test(s) checking this.

The leaf nodes are not necessarily aggregating (in case of `distinct`) so an empty sequence may be produced - the merger supposed to aggregate them - that's why this is needed.

For nested query stuff the merge runner becomes this [lambda](https://github.com/apache/druid/blob/d201ea0ece00cf4408b0f358d93071472c606c4e/processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java#L125-L132) (note: I don't know why I didn't placed this call there - just moved it)

example tests
* [testCountDistinctNonApproximateEmptySet](https://github.com/apache/druid/pull/14748/files#diff-fe60470a1d97fa758f70099a5ed49168cfe57f726dc6f08578df9b575b9a21a1R1953) is a sql level one 
* [testSummaryrowForEmptySubqueryInput](https://github.com/apache/druid/pull/14748/files#diff-df810dbf0150a412cc7b71f379267ecd7f1f1f4da906e961d7cbce718147d7b7R13004) as a runnertest

**Review by clintropolis — APPROVED** [review]

**Comment by kgyrtkirk** [comment]
The last test results have uncovered that `HAVING` clauses were not able to filter the summary row - because it was added after those were processed.

To avoid that issue I've moved the insertion of the optional summary row to be right before postprocessing is applied

**Review by clintropolis — APPROVED** [review]
